webapp/components/notifications/NotificationMenu/NotificationMenu.vue

```diff
@@ -46,24 +48,53 @@ export default {
           variables,
         })
         if (!(markAsRead && markAsRead.read === true)) return
-        this.notifications = this.notifications.map(n => {
-          return n.from.id === markAsRead.from.id ? markAsRead : n
+        this.displayedNotifications = this.displayedNotifications.map(n => {
+          return this.equalNotification(n, markAsRead) ? markAsRead : n
         })
       } catch (err) {
-        throw new Error(err)
+        this.$toast.error(err.message)
       }
     },
+    equalNotification(a, b) {
+      return a.from.id === b.from.id && a.createdAt === b.createdAt && a.reason === b.reason
+    },
   },
   computed: {
-    totalNotifications() {
-      return (this.notifications || []).length
+    notificationsCount() {
+      return (this.displayedNotifications || []).length
+    },
+    unreadNotificationsCount() {
+      let countUnread = 0
+      if (this.displayedNotifications) {
+        this.displayedNotifications.forEach(notification => {
+          if (!notification.read) countUnread++
+        })
```

Comment:
> [<img alt="roschaefer" height="40" width="40" align="left" src="https://avatars1.githubusercontent.com/u/38261864?s=88&v=4">](https://github.com/roschaefer) **Authored by [roschaefer](https://github.com/roschaefer)**
_<time datetime="2019-09-26T07:26:45Z" title="Thursday, September 26th 2019, 9:26:45 am +02:00">Sep 26, 2019</time>_

---
> **Outdated (history rewrite)** - original diff
---
```diff
@@ -29,13 +41,37 @@ export default {
   },
   data() {
     return {
+      displayedNotifications: [],
       notifications: [],
     }
   },
   props: {
     placement: { type: String },
   },
+  watch: {
+    notifications: {
+      immediate: true,
+      handler(lastQueriedNotifications) {
+        if (lastQueriedNotifications && lastQueriedNotifications.length > 0) {
+          lastQueriedNotifications.forEach(updatedListNotification => {
+            const sameNotification = this.displayedNotifications.find(oldListNotification => {
+              return this.equalNotification(oldListNotification, updatedListNotification)
+            })
+            if (!sameNotification) this.displayedNotifications.unshift(updatedListNotification)
+          })
+        }
+      },
+    },
+  },
```


@Tirokk you don't have to put this into a `watch` handler.

Vue-apollo has the `update` for graphql queries. We use it in a couple of other locations.  It's suitable for this task when you want to define **how** the `notifications` should be updated (in your case, you want to merge the new notifications with the old ones).

See here: https://vue-apollo.netlify.com/api/smart-query.html#options

We cannot use `fetchMore` as far as I know, because we're not using pagination here :thinking:

Reply:
> [<img alt="Tirokk" height="40" width="40" align="left" src="https://avatars1.githubusercontent.com/u/38261864?s=88&v=4">](https://github.com/Tirokk) **Authored by [Tirokk](https://github.com/Tirokk)**
_<time datetime="2019-09-26T08:50:17Z" title="Thursday, September 26th 2019, 10:50:17 am +02:00">Sep 26, 2019</time>_

---


That’s wonderful as well 😍 …

~~BUT: Now it's so perfect, so that I can't even test my code anymore on my own.
Because if I write myself a notification it updates immediately at sending the comment. 
😂🤣😅😰 TOP !!!~~

~~But I could use two browsers ???
NO: Does not work ether. Apollo is just to intelligent 😬😬😬~~

This is funny, but I made a mistake that always updated the notifications. Therefore I had the impression of magic … 💫

Reply:
> [<img alt="Tirokk" height="40" width="40" align="left" src="https://avatars1.githubusercontent.com/u/38261864?s=88&v=4">](https://github.com/Tirokk) **Authored by [Tirokk](https://github.com/Tirokk)**
_<time datetime="2019-09-26T08:51:21Z" title="Thursday, September 26th 2019, 10:51:21 am +02:00">Sep 26, 2019</time>_

---


Isn’t `refetch` enough?
What I use.
You mean we could save communication resources … ?

Reply:
> [<img alt="roschaefer" height="40" width="40" align="left" src="https://avatars1.githubusercontent.com/u/38261864?s=88&v=4">](https://github.com/roschaefer) **Authored by [roschaefer](https://github.com/roschaefer)**
_<time datetime="2019-09-26T12:50:10Z" title="Thursday, September 26th 2019, 2:50:10 pm +02:00">Sep 26, 2019</time>_

---


I was confused by `update` earlier: It triggers two times. Feel free to debug it with `console.log`. The first call comes when apollo reads from the cache. As soon as the new results from the server arrives, `update` is called again. It might be that apollo is so smart that it updates the cache if you mark a notification as `read`. :man_shrugging:

Reply:
> [<img alt="roschaefer" height="40" width="40" align="left" src="https://avatars1.githubusercontent.com/u/38261864?s=88&v=4">](https://github.com/roschaefer) **Authored by [roschaefer](https://github.com/roschaefer)**
_<time datetime="2019-09-26T12:51:20Z" title="Thursday, September 26th 2019, 2:51:20 pm +02:00">Sep 26, 2019</time>_

---


Maybe `refetch` also works. I think `refetch` only triggers once. We use it for pagination.